pkg/ivd/ivd_protected_entity_type_manager_util.go

```diff
@@ -0,0 +1,204 @@
+
+	var cnsVolumeCreateSpecList []cnstypes.CnsVolumeCreateSpec
+	cnsVolumeCreateSpec := cnstypes.CnsVolumeCreateSpec{
+		Name:        md.VirtualStorageObject.Config.Name,
+		VolumeType: string(cnstypes.CnsVolumeTypeBlock),
+		Datastores: dsList,
+		Metadata: cnstypes.CnsVolumeMetadata{
+			ContainerCluster: cnstypes.CnsContainerCluster{
+				ClusterType: string(cnstypes.CnsClusterTypeKubernetes), // hard coded for the moment
+				ClusterId:   clusterId,
+				VSphereUser: user,
+			},
+			EntityMetadata: metadataList,
+		},
+		BackingObjectDetails: &cnstypes.CnsBlockBackingDetails{
+			CnsBackingObjectDetails: cnstypes.CnsBackingObjectDetails{
+				CapacityInMb: md.VirtualStorageObject.Config.CapacityInMB,
+			},
+		},
+	}
+
+	cnsVolumeCreateSpecList = append(cnsVolumeCreateSpecList, cnsVolumeCreateSpec)
+	logger.Debugf("Provisioning volume using the spec: %v", cnsVolumeCreateSpec)
+
+	// provision volume using CNS API
+	cnsClient, err := cns.NewClient(ctx, client.Client)
+	if err != nil {
+		return "", err
+	}
+	createTask, err := cnsClient.CreateVolume(ctx, cnsVolumeCreateSpecList)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to create volume. Error: %+v", err)
+		return "", err
+	}
+	createTaskInfo, err := cns.GetTaskInfo(ctx, createTask)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to create volume. Error: %+v", err)
+		return "", err
+	}
+	createTaskResult, err := cns.GetTaskResult(ctx, createTaskInfo)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to create volume. Error: %+v", err)
+		return "", err
+	}
+	if createTaskResult == nil {
+		err := errors.New("Empty create task results")
+		logger.Error(err.Error())
+		return "", err
+	}
+	createVolumeOperationRes := createTaskResult.GetCnsVolumeOperationResult()
+	if createVolumeOperationRes.Fault != nil {
+		logger.Errorf("Failed to create volume: fault=%+v", createVolumeOperationRes.Fault)
+		return "", errors.New(createVolumeOperationRes.Fault.LocalizedMessage)
+	}
+
+	volumeId := createVolumeOperationRes.VolumeId.Id
+	logger.Infof("CNS volume, %v, created", volumeId)
+	return volumeId, nil
+}
+
+func filterLabelsFromMetadata(md Metadata, prefix string, logger logrus.FieldLogger) Metadata {
+	// prefix: cns.containerCluster
+	var kvsList []vim25types.KeyValue
+
+	logger.Debugf("labels of CNS voluem before filtering ones with certain prefix, %v: %v", prefix, md.ExtendedMetadata)
+
+	for _, label := range md.ExtendedMetadata {
+		if !strings.HasPrefix(label.Key, prefix) {
+			kvsList = append(kvsList, vim25types.KeyValue {
+				Key: label.Key,
+				Value: label.Value,
+			})
+		}
+	}
+	md.ExtendedMetadata = kvsList
+
+	logger.Debugf("labels of CNS voluem after filtering ones with certain prefix, %v: %v", prefix, md.ExtendedMetadata)
+
+	return md
+}
+
+func CreateCnsVolumeInCluster(ctx context.Context, client *govmomi.Client, md Metadata, logger logrus.FieldLogger) (string, error) {
```

Comment:
Can we have this return an ID rather than a string.  You just convert it to an ID immediately anyhow

Reply:
Yes, we can.